### ISSUE TITLE
[ext_proc] STREAMED request body mode could result in duplicate requests sent to endpoint

### ISSUE THREAD

**Issue by johnsiu** [envoyproxy/envoy#27216]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: *STREAMED request body mode could result in duplicate request frames sent to endpoint*

*Description*:
> Request frames sent to endpoint should not be duplicated.

*Repro steps*:
> run envoy with config as follows:
```
admin:
    address:
      socket_address: { address: 127.0.0.1, port_value: 9901 }
  
  static_resources:
    listeners:
    - name: listener_0
      address:
        socket_address: { address: 127.0.0.1, port_value: 10000 }
      filter_chains:
      - filters:
        - name: envoy.filters.network.http_connection_manager
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
            stat_prefix: ingress_http
            codec_type: AUTO
            route_config:
              name: local_route
              virtual_hosts:
              - name: local_service
                domains: ["*"]
                routes:
                - match: { prefix: "/" }
                  route: { cluster: some_service }
            http_filters:
            - name: envoy.filters.http.ext_proc
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.http.ext_proc.v3.ExternalProcessor
                failure_mode_allow: false
                processing_mode:
                  request_header_mode: "SEND"
                  response_header_mode: "SKIP"
                  request_body_mode: "STREAMED"
                  response_body_mode: "NONE"
                  request_trailer_mode: "SKIP"
                  response_trailer_mode: "SKIP"
                grpc_service:
                  google_grpc:
                    target_uri: localhost:<portA>
                    stat_prefix: testextproc
            - name: envoy.filters.http.router
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
    clusters:
    - name: some_service
      connect_timeout: 10s
      type: STRICT_DNS
      lb_policy: ROUND_ROBIN
      http2_protocol_options: {}
      transport_socket:
        name: tls
        typed_config:
          "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext"
          common_tls_context:
            tls_certificates:
            - certificate_chain:
                filename: <some cert>
              private_key:
                filename: <some key>
      load_assignment:
        cluster_name: some_service
        endpoints:
        - lb_endpoints:
          - endpoint:
              address:
                socket_address: 
                  address: <some endpoint>
                  port_value: <portB>
      health_checks:
        timeout: 10s
        interval: 10s
        unhealthy_threshold: 2
        healthy_threshold: 2
        grpc_health_check: {}
``` 
- Run an external processor grpc server on `localhost:<portA>`
- The external processor simply returns empty responses for `request_headers` and `request_body` to indicate no-op.
- Run any dummy grpc server on `<some endpoint>:<portB>`. It should support 1 unary RPC and 1 client streaming RPC.
- Run a moderate throughput of unary grpc calls to `127.0.0.1:10000`. 
- Observe that the external processor grpc server receives 1 RequestHeaders ProcessingRequest and 1 RequestBody ProcessingRequest for each RPC call as expected
- Observe that the dummy grpc server could non-deterministically receive 1 or more request for each RPC call. Since this is a unary RPC, having received more than 1 request violated the protocol and resulted in `Too many requests` error from the dummy grpc server.
- repeat the above test using client streaming RPC.
- Observe that the external processor grpc server receives 1 RequestHeaders ProcessingRequest and `n` RequestBody ProcessingRequest that matches the number of request frames sent by the caller for each RPC call as expected
- Observe that the dummy grpc server could non-deterministically receive duplicated request frames for each RPC call.

**Comment by RyanTheOptimist** [comment]
cc: @adisuissa

**Comment by adisuissa** [comment]
cc @yanjunxiang-google can you PTAL?

**Comment by yanjunxiang-google** [comment]
I think we need a cluster to pointing to  localhost:<portA>.

@tyxia @stevenzzzz

**Comment by stevenzzzz** [comment]
> I think we need a cluster to pointing to localhost:.
> 
> @tyxia @stevenzzzz

more details why you think so? 

I think we need more details to diagnosis the real issue here. 

OTOH, can we improve our fuzz to mimic this behavior described by [johnsiu](https://github.com/johnsiu)?

**Comment by tyxia** [comment]
> I think we need a cluster to pointing to localhost:.
> 
> @tyxia @stevenzzzz

Could you please elaborate a bit? Are you referring to the cluster for grpc? Only envoy grpc needs the cluster I think.  Config here is showing the google_grpc

**Comment by johnsiu** [comment]
I can confirm that the bug has been fixed in release 1.28. Thank you.

**Comment by yanjunxiang-google** [comment]
Thanks for the update! There are some cleanups for ext_proc filter STREAMED mode in release 1.28.
